scipy/signal/_spectral_py.py

```diff
@@ -134,22 +150,120 @@ def lombscargle(x,
+        normalize = OUTPUT_NORMALIZE if normalize else OUTPUT_POWER
     else:
-        pgram = _lombscargle(x, y, freqs)
-
-    if normalize:
-        pgram *= 2 / np.dot(y, y)
+        # if neither a bool or str
+        if not isinstance(normalize, str):
+            raise TypeError("Normalize type must a bool or str.")
+
+    if normalize not in [OUTPUT_POWER, OUTPUT_NORMALIZE, OUTPUT_AMPLITUDE]:
+        raise ValueError(
+            "Normalize must be: "
+            f"False (or '{OUTPUT_POWER}'), "
+            f"True (or '{OUTPUT_NORMALIZE}'), "
+            f"or '{OUTPUT_AMPLITUDE}'."
+        )
 
+    # convert inputs to contiguous arrays with high precision
+    x = np.ascontiguousarray(x, dtype=np.float64)
+    y = np.ascontiguousarray(y, dtype=np.float64)
+    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
+    weights = np.ascontiguousarray(weights, dtype=np.float64)
+
+    # weight vector must sum to 1
+    weights = weights / weights.sum()
+
+    # pre-allocate arrays for intermediate and final calculations
+    coswt = np.empty_like(x)
+    sinwt = np.empty_like(x)
+    wcoswt = np.empty_like(x)
+    wsinwt = np.empty_like(x)
+    pgram = np.empty_like(freqs)
+    # store a and b so that phase can be calculated outside loop, if necessary
+    a = np.empty_like(freqs)
+    b = np.empty_like(freqs)
+
+    # calculate the single sum that does not depend on the frequency
+    Y_sum = (weights * y).sum()
+
+    # loop over the frequencies
+    for i in range(freqs.shape[0]):
+        coswt[:] = np.cos(freqs[i] * x)
+        sinwt[:] = np.sin(freqs[i] * x)
+        wcoswt[:] = weights * coswt
+        wsinwt[:] = weights * sinwt
+        C_sum = wcoswt.sum()
+        S_sum = wsinwt.sum()
+
+        YC_hat = (y * wcoswt).sum()
+        YS_hat = (y * wsinwt).sum()
+        CC_hat = (wcoswt * coswt).sum()
+        SS_hat = 1 - CC_hat  # trig identity: S^2 = 1 - C^2
+        CS_hat = (wcoswt * sinwt).sum()
+
+        YC = YC_hat - Y_sum * C_sum
+        YS = YS_hat - Y_sum * S_sum
+        CC = CC_hat - C_sum * C_sum
+        SS = SS_hat - S_sum * S_sum
+        CS = CS_hat - C_sum * S_sum
+        D = CC * SS - CS * CS
+
+        # where: y(w) = a*cos(w) + b*sin(w) + offset
+        a[i] = (YC * SS - YS * CS) / D
+        b[i] = (YS * CC - YC * CS) / D
+        # offset = Y_sum - a * C_sum - b * S_sum  # not useful to return
+
+        # store final value as power in (y units) ^ 2
+        pgram[i] = 2.0 * (a[i] * YC + b[i] * YS)
+
+    if normalize == OUTPUT_NORMALIZE:
+        # return the normalized power (current frequency wrt the entire signal)
+        YY_hat = (weights * y * y).sum()
+        YY: float = YY_hat - Y_sum * Y_sum
+        pgram /= 2.0 * YY
+        return pgram
+
+    elif normalize == OUTPUT_AMPLITUDE:
+        # return the complex representation of the amplitude and phase
+        phase = np.arctan2(b, a)  # radians
+        pgram = np.sqrt(pgram) * (np.cos(phase) + 1j * np.sin(phase))
+        return pgram
```

Comment:
My hunch is that 
```python
pgram = a + 1j*b
```
would produce the same result (while being faster). Could be easily verified by a unit test.

Reply:
I wish it was, but I found there are edge cases where this doesn't work. I'll have to dig back through my notes and simulations to find some examples.

Reply:
Okay, I found where the simple `pgram = a + 1j * b` breaks down: near 0. I also verified, using a symbolic math library, that no, the two equations are not equivalent.

The amplitude starts increasing exponentially, while the phase is still correct. See these two plots for comparing `correct formula (blue)` with `simple, but incorrect formula (orange)`.

![amplitude](https://github.com/user-attachments/assets/6de44bed-c537-4b2f-8a5b-11a376c35690)

![phase](https://github.com/user-attachments/assets/511a4d53-a199-465d-b90c-d96e6a05498b)

Reply:
Hmm, it looks a bit strange that those plots do not show a systematic deviation for all frequencies. It smells a lot like a numeric problem. I toyed a bit with an alternative implementation, which utilized [lstsq] directly, which had numeric instabilities near `f=0` as well (which could be illustrated by looking at the condition number). What was noticeable that the estimated offset also grew near `f=0`, just like in your plot (which is not really surprising, since all spectral lobes have some width in frequency). Do you happen to know what the true parameters of that dataset are? 

That said, it is probably best to keep the existing implementation for the time being.



[lstsq]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lstsq.html

Reply:
Yeah. Before the first commit, I had spent a little time refactoring the equations to make the code not only a little easier to follow while also making the transforms to each of the 3 possible output units neater as well. (I still think the "default" units are nonsense and not a form I've ever used in my career, but backwards compatibility and all).

When I was playing around 2 days ago with the complex output, I was able to get other forms that didn't exhibit this issue, but they are pretty gnarly. Although it looks kinda dumb to do the two-step to phase and back to a complex unit circle, it's really the amplitude that's causing the problem (as we can see). `YC` and `YS` tend towards zero as `f` approaches 0, which perfectly counteract the explosion of `a` and `b` (caused by `D` going towards zero too) in the amplitude (`sqrt(2*(a*YC + b*YS))`).


Here's the code I used:

    nin = 10000
    nout = 1000

    p = 0.7 
    np.random.seed(2353425)
    r = np.random.rand(nin)
    t = np.linspace(0.01*np.pi, 10.*np.pi, nin)[r >= p]
    w = np.linspace(0.01, 5., nout) 

    w_i = 100
    phi = 0 
    A = 2
    off = 1
    y = A*np.cos(w[w_i]*t + phi) + off
    
    # modified to also output `pgram2 = a + 1j * b`
    ls_gen_complex, ls_gen_complex_alt = lombscargle(t, y, w, normalize = 'amplitude')

    ax = sns.lineplot(x=w,y=np.abs(ls_gen_complex))
    sns.lineplot(x=w,y=np.abs(ls_gen_complex_alt), ax=ax)
    ax.set_yscale('log')
    ax.set_title('amplitude')
    ax.set_ylabel('amplitude (units)')
    ax.set_xlabel('frequency (rad/sec)')

    ax = sns.lineplot(x=w,y=np.angle(ls_gen_complex))
    sns.lineplot(x=w,y=np.angle(ls_gen_complex_alt), ax=ax)
    ax.set_title('angle')
    ax.set_ylabel('angle (radians)')
    ax.set_xlabel('frequency (rad/sec)')

Reply:
After looking at your explanations and doing a little bit more of reading, I understand now that the Lombscargle method may not be interpreted as an alternative formulation of the  least squares minimization of $y_i = a cos ⁡ ( ω _i ) + b cos ⁡ ( ω t_i ) + o_i$. 

Regarding the "critical limit": Formulating generic criteria for input data when the Lombscargle method becomes unreliable is a larger feat, as far as I understand. It heavily depends on the underlying noise model(s) and the desired accuracy. Even for the case of the standard linear least squares (where the condition number provides a sort of quality measure of the estimate), it is not straightforward. 

Regarding the ``f < (2*pi)/(t_max-t_min)`` heuristic: My guts think that  that condition might be too fuzzy to be useful—but I am not able to pinpoint the problem. Determining some sort of standard deviation of the average sample rate (or some other statistic) would appear more plausible to me. Some literature reference in that regard might helpful. 

For this PR, I think it would be sufficient to document that the quality of the Lombscargle method heavily depends on the temporal distribution of the input samples and perhaps that there is no direct equivalent to the Nyquist frequency and the frequency resolution being inversely proportional to the signal duration.

Regarding the documentation: I think [VanderPlas's paper] might be a useful reference, since it discusses the difference between standard FFT periodogram and the Lombscargle method from an application perspective.

[VanderPlas's paper]: https://iopscience.iop.org/article/10.3847/1538-4365/aab766

Reply:
Yeah, there are a ton of gotchas and not straightforward heuristics with non-uniform sampling in general.  See also many papers on trying to derive robust formulas for a "pseudo Nyquist limit" (this is a -fun- rabbit hole, and has provided me with hours of debated).

For now, I'll remove the check and add the reference.

Reply:
Done.